# Olist Project - Sellers Data Ingestion & Quality Check

**Dataset:** `olist_sellers_dataset.csv`  
**Purpose:** Load the sellers dataset with Pandas, inspect its structure, and validate key data-quality conditions before using it in seller-level and geographic analysis.

**Workflow:** CSV → Pandas DataFrame → structure check → null/duplicate/key checks → location sanity checks → findings

## 1. Setup

Import Pandas, which we use for CSV ingestion and data-quality checks.

In [11]:
import pandas as pd

## 2. Load the sellers dataset

`pd.read_csv()` reads the CSV file and creates a Pandas DataFrame.

In [12]:
sellers = pd.read_csv("../data/csv/olist_sellers_dataset.csv")

## 3. Initial inspection

These checks give a quick overview of dataset size, sample rows, column names, data types, and non-null counts.

In [13]:
print("Shape:", sellers.shape)
display(sellers.head())
print("Columns:", sellers.columns.tolist())
sellers.info()

Shape: (3095, 4)


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Columns: ['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']
<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


### Initial structure

Observed dataset structure:

- **3,095 rows**
- **4 columns**
- Columns:
  - `seller_id`
  - `seller_zip_code_prefix`
  - `seller_city`
  - `seller_state`

**Grain candidate:** one row per seller.

## 4. Data Quality Checks

### 4.1 Null values

`isnull().sum()` counts missing values column by column.

In [14]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

**Result:** No null values were found in the sellers dataset.

### 4.2 Full duplicate rows

`duplicated().sum()` checks whether any rows are completely identical across all columns.

In [15]:
sellers.duplicated().sum()

np.int64(0)

**Result:** `0` full duplicate rows.

### 4.3 Seller ID uniqueness

`seller_id` should uniquely identify a seller.  
Comparing total row count with distinct `seller_id` count verifies whether it behaves like a primary key.

In [16]:
print("Total rows:", len(sellers))
print("Unique seller IDs:", sellers["seller_id"].nunique())

Total rows: 3095
Unique seller IDs: 3095


**Result:**

- Total rows: **3,095**
- Unique `seller_id`: **3,095**

Therefore, `seller_id` is unique in this dataset and is a strong primary-key candidate.

### 4.4 ZIP-code prefix uniqueness

ZIP-code prefix is a geographic attribute, not a seller identifier.  
Multiple sellers can therefore legitimately share the same ZIP-code prefix.

In [17]:
sellers["seller_zip_code_prefix"].nunique()

2246

**Result:** **2,246** distinct ZIP-code prefixes across **3,095** sellers.

This is expected and does not indicate duplication.

### 4.5 Blank city and state values

A field can be non-null but still contain an empty string or only whitespace.  
So blank-string checks are performed separately from null checks.

The logic below:

1. selects the column,
2. converts values to string,
3. removes leading/trailing whitespace with `.str.strip()`,
4. checks whether the cleaned value equals `""`,
5. counts the `True` results.

In [18]:
print(
    "Blank cities:",
    sellers["seller_city"].astype(str).str.strip().eq("").sum()
)

print(
    "Blank states:",
    sellers["seller_state"].astype(str).str.strip().eq("").sum()
)

Blank cities: 0
Blank states: 0


**Result:**

- Blank city values: **0**
- Blank state values: **0**

So the checked location fields contain neither nulls nor blank strings.

### 4.6 State code sanity check

Brazilian state codes are represented as two-character abbreviations in this dataset.  
This check returns any rows where `seller_state` does not have the expected length.

In [19]:
sellers[sellers["seller_state"].astype(str).str.strip().str.len() != 2]

,seller_id,seller_zip_code_prefix,seller_city,seller_state


**Result:** No invalid state-code lengths were found.

### 4.7 State distribution

This is a light sanity check to inspect the observed state codes and seller counts.

In [20]:
sellers["seller_state"].value_counts().sort_index()

seller_state
AC       1
AM       1
BA      19
CE      13
DF      30
ES      23
GO      40
MA       1
MG     244
MS       5
MT       4
PA       1
PB       6
PE       9
PI       1
PR     349
RJ     171
RN       5
RO       2
RS     129
SC     190
SE       2
SP    1849
Name: count, dtype: int64

## 5. MySQL Preparation

The sellers dataset passed the initial Pandas quality checks, so the next step is to load it into MySQL.

For this project, the table schema is created explicitly in SQL before using `to_sql()`. This keeps control over data types and the primary key instead of letting Pandas/SQLAlchemy infer the table structure automatically.

The corresponding SQL file is:

`sql/09_create_sellers_table.sql`

```sql
USE olist_ecommerce;

CREATE TABLE sellers (
    seller_id VARCHAR(32) PRIMARY KEY,
    seller_zip_code_prefix INT,
    seller_city VARCHAR(100),
    seller_state CHAR(2)
);
```

**Why create the table first?**

- `seller_id` can be defined explicitly as the primary key.
- Column data types stay under our control.
- The database schema remains documented in the SQL folder.
- `to_sql(..., if_exists="append")` then performs only the data-load step.

## 6. Connect Python to MySQL

SQLAlchemy is used to connect the Pandas workflow to the existing `olist_ecommerce` MySQL database.

`getpass()` keeps the password out of the notebook output, while `quote_plus()` safely encodes special characters that may appear in the password.

In [22]:
from sqlalchemy import create_engine, text
from getpass import getpass
from urllib.parse import quote_plus


In [23]:
mysql_user = "root"
mysql_password = getpass("MySQL password: ")
mysql_password = quote_plus(mysql_password)

engine = create_engine(
    f"mysql+pymysql://{mysql_user}:{mysql_password}@localhost:3306/olist_ecommerce?charset=utf8mb4"
)


### 6.1 Validate the database connection

Before loading data, confirm that the engine is connected to the intended database.

In [24]:
with engine.connect() as connection:
    database = connection.execute(text("SELECT DATABASE();")).scalar()
    print(database)

# olist_ecommerce


olist_ecommerce


**Result:** The connection returned `olist_ecommerce`, confirming that Python is connected to the correct database.

## 7. Load the Sellers DataFrame into MySQL

Because the `sellers` table was already created in MySQL, `if_exists="append"` is used so Pandas inserts rows into the existing schema rather than replacing it.

Key parameters:

- `name="sellers"` → target MySQL table
- `if_exists="append"` → insert into the existing table
- `index=False` → do not create an extra Pandas index column
- `chunksize=1000` → send rows in batches
- `method="multi"` → group multiple rows into each INSERT operation

In [25]:
sellers.to_sql(
    name="sellers",
    con=engine,
    if_exists="append",
    index=False,
    chunksize=1000,
    method="multi"
)

3095

**Result:** `3,095` rows were inserted into the MySQL `sellers` table.

## 8. Validate the MySQL Load

The database result is compared with the Pandas QC result to verify that the load did not lose or duplicate records.

In [26]:
pd.read_sql(
    """
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT seller_id) AS unique_seller_ids,
        COUNT(DISTINCT seller_zip_code_prefix) AS unique_zip_prefixes
    FROM sellers;
    """,
    con=engine
)

,total_rows,unique_seller_ids,unique_zip_prefixes
0,3095,3095,2246


### Validation Result

| Metric | Pandas | MySQL |
|---|---:|---:|
| Total rows | 3,095 | 3,095 |
| Unique `seller_id` | 3,095 | 3,095 |
| Unique ZIP-code prefixes | 2,246 | 2,246 |

The Pandas and MySQL counts match exactly.

**Validation:** The sellers dataset was loaded without row loss or row multiplication.

## 9. Final Sellers QC & Load Summary

| Check | Result |
|---|---:|
| Total rows | 3,095 |
| Unique `seller_id` | 3,095 |
| Unique ZIP-code prefixes | 2,246 |
| Null values | 0 |
| Full duplicate rows | 0 |
| Blank city values | 0 |
| Blank state values | 0 |
| Invalid state-code length | 0 |
| Rows loaded to MySQL | 3,095 |
| Pandas ↔ MySQL row-count match | Yes |
| Pandas ↔ MySQL unique seller-count match | Yes |

### Interpretation

- **Grain:** one row represents one seller.
- `seller_id` uniquely identifies each seller and is used as the MySQL primary key.
- ZIP-code prefix is not unique, which is expected because multiple sellers may operate in the same geographic area.
- No missing, blank, duplicate, or obvious state-format issues were found.
- MySQL validation matched the Pandas source counts exactly.
- The sellers table is now ready for seller-level, geographic, and later Delivery & Customer Experience analysis.

## 10. Python / Pandas / SQLAlchemy Concepts Used

| Concept | What it does | SQL-style analogy |
|---|---|---|
| `pd.read_csv()` | CSV → DataFrame | Import / load data |
| `.shape` | Row and column count | `COUNT(*)` + schema size |
| `.head()` | Shows first rows | `LIMIT` |
| `.columns.tolist()` | Lists column names | Inspect schema |
| `.info()` | Data types + non-null counts | `DESCRIBE` + null summary |
| `.isnull().sum()` | Counts nulls by column | `SUM(col IS NULL)` |
| `.duplicated().sum()` | Counts full duplicate rows | Duplicate row check |
| `.nunique()` | Counts distinct values | `COUNT(DISTINCT ...)` |
| `.astype(str)` | Converts values to string | `CAST(... AS CHAR)` |
| `.str.strip()` | Removes leading/trailing whitespace | `TRIM()` |
| `.eq("")` | Tests equality with an empty string | `= ''` |
| `.str.len()` | Returns string length | `CHAR_LENGTH()` |
| `.value_counts()` | Counts frequency of values | `GROUP BY + COUNT(*)` |
| `.sort_index()` | Sorts by index/value label | `ORDER BY` |
| `create_engine()` | Creates a SQLAlchemy database connection engine | Database connection |
| `getpass()` | Reads the password without displaying it | Credential handling |
| `quote_plus()` | Safely encodes special URL characters | Connection-string preparation |
| `.to_sql()` | DataFrame → SQL table | `INSERT` / bulk load |
| `pd.read_sql()` | SQL query result → DataFrame | `SELECT` result |
| `if_exists="append"` | Inserts into an existing table | `INSERT INTO` |
| `chunksize=1000` | Loads data in batches | Batched inserts |
| `method="multi"` | Inserts multiple rows per statement | Multi-row `INSERT` |

## 11. Next Step

Document the validated sellers findings in:

`sql/10_data_quality_checks.sql`

Then save the notebook and include both the notebook and SQL updates in the next GitHub commit.

**Completed workflow:**  
CSV → Pandas ingestion → Python QC → explicit MySQL schema → Python-to-MySQL load → database validation ✅